neuraxle/hyperparams/space.py

```diff
@@ -226,6 +228,160 @@ class HyperparameterSamples(RecursiveDict):
+        ...OrderedDict([('AddFeatures__PCA', dict([('copy', True), ('iterated_power', 'auto')]))])
+        """
+        grouped_hyper_params = OrderedDict()
+        for steps_and_hps, hyper_param_val in flat_steps_hps:
+            *steps, hyper_param_key = steps_and_hps.split(self._separator)
+            grouped_hyper_params.setdefault(f"{self._separator}".join(steps), dict()).update({
+                hyper_param_key: hyper_param_val})
+
+        return grouped_hyper_params
+
+    def _convert_to_compressed_format(self, group_by_step: 'OrderedDict[str, dict]') -> 'List[dict]':
+        """Converts grouped hyper params to Compressed format."""
+        compressed = []
+        for steps in group_by_step:
+            *prev_steps, current_step = steps.split(self._separator)
+            hps: dict = group_by_step[steps]
+
+            compressed.append(CompressedHyperparameter(step_name=current_step, hyperparams=hps,
+                                                       ancestor_steps=prev_steps
+                                                       if not self._trim_parents else None).__dict__)
+        return compressed
+
+    def _convert(self) -> 'List[dict]':
+        grouped_result: OrderedDict[str, dict] = self._group_hps_by_step(flat_steps_hps=self._flat_hps.items())
+        return self._convert_to_compressed_format(group_by_step=grouped_result)
+
+    def wildcards(self) -> OrderedDict:
+        return self._convert_to_wild_cards()
+
+    def _generate_reverse_suffix_strings(self, string_array: List[str]):
+        """
+        Generates reverse suffix set of strings for an given input array of strings
+        :param string_array:
+        :return:
+
+        >>>ip_array = ["a","b","c"]
+        >>>self._separator = '__'
+        >>>self._generate_reverse_suffix_strings(ip_array)
+        ...{'a__b__c', 'b__c', 'c'}
+        """
+        suffix_strings = set()
+        for idx in range(1, len(string_array) + 1):
+            suffix_strings.add(f"{self._separator}".join(string_array[-idx:]))
+        return suffix_strings
+
+    def _generate_string_pairs(self, string_array: List[str]):
+        """
+        Generates set of string pairs tuple
+        :param string_array:
+        :return:
+
+        >>>temp_array = ["a","b","c"]
+        >>>self._generate_string_pairs(temp_array)
+        ...{('a','b'),('a','c')}
+        """
+        string_pairs = set()
+        y = string_array[-1]
+        for x in string_array[:-1]:
+            string_pairs.add((x, y))
+        return string_pairs
+
+    def _can_be_pruned_further(self, unique_str: List[str], string_pairs_freq: Counter) -> bool:
+        """
+        Checks if unique string can be pruned or not
+        """
+        return string_pairs_freq[(unique_str[0], unique_str[-1])] <= 1
+
+    def _convert_to_wild_cards(self) -> OrderedDict:
+        """
+        Method compresses the self._flat_hps into wildcard format
+        """
```

Comment:
It seems that this method would need to be splitted into more different private methods to make all of them simpler to read. Reference: https://softwareengineering.stackexchange.com/a/195992

Reply:
You could do more unit tests with this to test each method separately if that helps you code faster. Often it does help people code faster because it helps debugging.

Reply:
Thanks for the comment. I will see through to split these into multiple methods.

Reply:
Also as per the practice, it is not recommended to unit test the private methods. But if that is expected then i will also write tests for the private methods aswell.